pages/directory/index.tsx

```diff
@@ -8,6 +9,56 @@ import createAxiosInstance from 'client/lib/axios';
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 const Directory: NextPage<{ users: any[]; error: string }> = props => {
   const { users, error } = props;
+  const lastCardRef = React.useRef(null);
+  const [userData, setUserData] = React.useState(users);
+  const [currentPage, setCurrentPage] = React.useState(1);
+  const [totalPages, setTotalPages] = React.useState(999);
+  const [isLoading, setIsLoading] = React.useState(false);
+
+  async function fetchUsers(page: number) {
+    try {
+      const response = await axios.get(`/api/users?page=${page}`);
+      return {
+        page: response.data.page,
+        total: response.data.totalPages,
+        users: response.data.users,
+      };
+    } catch (err) {
+      console.error(err);
+    }
+  }
+
+  React.useEffect(() => {
+    const intersectionOptions = {
+      // root default = document's viewport
+      // in the future offset should be tied to the card height of the last row
+      rootMargin: '300px',
+      threshold: 0,
+    };
+
+    const lastCardObserver = new IntersectionObserver(async ([lastCard]) => {
+      // guard clause, should intersect only once
+      if (!lastCard.isIntersecting) return;
+      lastCardObserver.disconnect();
+      setIsLoading(true);
+      try {
+        // stop loading users when at max pages
+        if (currentPage + 1 > totalPages) {
+          setIsLoading(false);
+          return;
+        }
+        const { page, total, users } = await fetchUsers(currentPage + 1);
+        setCurrentPage(page);
+        setTotalPages(total);
+        setUserData([...userData, ...users]);
```

Comment:
using `totalPages` state here to avoid an additional api call, I understand this could be handled by calling the api and if a 400 badRequest happens then we use that to determine the max pages that were hit. But this way after the first request we know the totalPages to look for and when those are hit then we can prevent any additional api calls